File: designer/src/com/android/tools/idea/uibuilder/handlers/relative/RelativeLayoutHandler.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.uibuilder.handlers.relative;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import com.android.tools.idea.uibuilder.api.*;
import com.android.tools.idea.uibuilder.graphics.NlGraphics;
import com.android.tools.idea.uibuilder.model.AndroidCoordinate;
import com.android.tools.idea.uibuilder.model.NlComponent;
import com.android.tools.idea.uibuilder.model.SegmentType;
import com.android.tools.idea.uibuilder.model.TextDirection;
import com.android.tools.idea.uibuilder.surface.ScreenView;
import com.google.common.collect.Lists;

import java.awt.*;
import java.util.List;

/** Handler for the {@code <RelativeLayout>} layout */
public class RelativeLayoutHandler extends ViewGroupHandler {
  @Override
  public boolean paintConstraints(@NotNull ScreenView screenView, @NotNull Graphics2D graphics, @NotNull NlComponent component) {
    NlGraphics g = new NlGraphics(graphics, screenView);
    Iterable<NlComponent> iterable = component.getChildren();
    List<NlComponent> children = Lists.newArrayList(iterable);
    ConstraintPainter.paintSelectionFeedback(g, component, children, true, TextDirection.LEFT_TO_RIGHT);
    return false;
  }

  @Override
  @Nullable
  public DragHandler createDragHandler(@NotNull ViewEditor editor,
                                       @NotNull NlComponent layout,
                                       @NotNull List<NlComponent> components,
                                       @NotNull DragType type) {
    final RelativeDragHandler moveHandler = new RelativeDragHandler(editor, layout, components);
    return new DragHandler(editor, this, layout, components, type) {
      @Nullable
      @Override
      public String update(@AndroidCoordinate int x, @AndroidCoordinate int y, int modifiers) {
        super.update(x, y, modifiers);
        NlComponent primary = components.get(0);
        int deltaX = lastX - startX;
        int deltaY = lastY - startY;
        moveHandler.updateMove(primary, deltaX, deltaY, modifiers);

        return null;
      }

      @Override
      public void paint(@NotNull NlGraphics graphics) {
        GuidelinePainter.paint(graphics, moveHandler);
      }

      @Override
      public void commit(@AndroidCoordinate int x, @AndroidCoordinate int y, int modifiers) {
        moveHandler.removeCycles();

        NlComponent previous = null;
        for (NlComponent component : components) {
          if (previous == null) {
            moveHandler.applyConstraints(component);
          } else {
            // Arrange the nodes next to each other, depending on which
            // edge we are attaching to. For example, if attaching to the
            // top edge, arrange the subsequent nodes in a column below it.
            //
            // TODO: Try to do something smarter here where we detect
            // constraints between the dragged edges, and we preserve these.
            // We have to do this carefully though because if the
            // constraints go through some other nodes not part of the
            // selection, this doesn't work right, and you might be
            // dragging several connected components, which we'd then
            // need to stitch together such that they are all visible.

            moveHandler.attachPrevious(previous, component);
          }
          previous = component;
        }
      }
    };
  }

  @Override
  @Nullable
  public ResizeHandler createResizeHandler(@NotNull ViewEditor editor,
                                           @NotNull NlComponent component,
                                           @Nullable SegmentType horizontalEdgeType,
                                           @Nullable SegmentType verticalEdgeType) {
    NlComponent parent = component.getParent();
    if (parent == null) {
      return null;
    }
    final RelativeResizeHandler resizeHandler = new RelativeResizeHandler(editor, parent, component, horizontalEdgeType, verticalEdgeType);

    return new ResizeHandler(editor, this, component, horizontalEdgeType, verticalEdgeType) {
      @Nullable
      @Override
      public String update(@AndroidCoordinate int x,
                           @AndroidCoordinate int y,
                           int modifiers,
                           @NotNull @AndroidCoordinate Rectangle newBounds) {
        super.update(x, y, modifiers, newBounds);
        resizeHandler.updateResize(component, newBounds, modifiers);
        return null;
      }

      @Override
      public void commit(@AndroidCoordinate int px,
                         @AndroidCoordinate int py,
                         int modifiers,
                         @NotNull @AndroidCoordinate Rectangle newBounds) {
        resizeHandler.removeCycles();
        resizeHandler.applyConstraints(component);
      }

      @Override
      public void paint(@NotNull NlGraphics graphics) {
        GuidelinePainter.paint(graphics, resizeHandler);
      }
    };
  }
}

File: designer/src/com/android/tools/idea/uibuilder/handlers/relative/RelativeLayoutHandler.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.uibuilder.handlers.relative

import com.android.SdkConstants
import com.android.tools.idea.common.api.DragType
import com.android.tools.idea.common.api.InsertType
import com.android.tools.idea.common.model.NlAttributesHolder
import com.android.tools.idea.common.model.NlComponent
import com.android.tools.idea.common.scene.SceneComponent
import com.android.tools.idea.common.scene.SceneInteraction
import com.android.tools.idea.common.scene.target.AnchorTarget
import com.android.tools.idea.common.scene.target.Target
import com.android.tools.idea.uibuilder.api.DragHandler
import com.android.tools.idea.uibuilder.api.ViewEditor
import com.android.tools.idea.uibuilder.api.ViewGroupHandler
import com.android.tools.idea.uibuilder.api.actions.ToggleAutoConnectAction
import com.android.tools.idea.uibuilder.api.actions.ViewAction
import com.android.tools.idea.uibuilder.api.actions.ViewActionUtils.getToggleSizeActions
import com.android.tools.idea.uibuilder.api.actions.ViewActionUtils.getViewOptionsAction
import com.android.tools.idea.uibuilder.handlers.relative.targets.RELATIVE_LAYOUT_ATTRIBUTES
import com.android.tools.idea.uibuilder.handlers.relative.targets.RelativeAnchorTarget
import com.android.tools.idea.uibuilder.handlers.relative.targets.RelativeDragTarget
import com.android.tools.idea.uibuilder.handlers.relative.targets.RelativeParentTarget
import com.android.tools.idea.uibuilder.handlers.relative.targets.RelativeResizeTarget
import com.android.tools.idea.uibuilder.handlers.relative.targets.RelativeWidgetTarget
import com.android.tools.idea.uibuilder.model.getBaseline
import com.android.tools.idea.uibuilder.scene.target.ResizeBaseTarget
import com.android.tools.idea.uibuilder.surface.ScreenView
import com.google.common.collect.ImmutableList

/**
 * Handler of New Target Architecture for the `<RelativeLayout>` layout
 *
 * TODO:
 * - Don't allow selecting multiple widgets.
 */
class RelativeLayoutHandler : ViewGroupHandler() {

  override fun createDragHandler(editor: ViewEditor, layout: SceneComponent, components: List<NlComponent>, type: DragType): DragHandler? {
    if (layout.drawWidth == 0 || layout.drawHeight == 0) {
      return null
    }
    return RelativeDragHandler(editor, this, layout, components, type)
  }

  override fun onChildRemoved(
    layout: NlComponent,
    newChild: NlComponent,
    insertType: InsertType
  ) {
    RELATIVE_LAYOUT_ATTRIBUTES.forEach { newChild.removeAndroidAttribute(it) }
  }

  override fun cleanUpAttributes(component: NlComponent, attributes: NlAttributesHolder) {
    with (attributes) {
      if (getAndroidAttribute(SdkConstants.ATTR_LAYOUT_CENTER_HORIZONTAL) == SdkConstants.VALUE_TRUE &&
          getAndroidAttribute(SdkConstants.ATTR_LAYOUT_CENTER_VERTICAL) == SdkConstants.VALUE_TRUE) {
        removeAndroidAttribute(SdkConstants.ATTR_LAYOUT_CENTER_HORIZONTAL)
        removeAndroidAttribute(SdkConstants.ATTR_LAYOUT_CENTER_VERTICAL)
        setAndroidAttribute(SdkConstants.ATTR_LAYOUT_CENTER_IN_PARENT, SdkConstants.VALUE_TRUE)
      }

      val margin = getAndroidAttribute(SdkConstants.ATTR_LAYOUT_MARGIN)
      if (margin != null) {
        MARGINS_ATTRS.forEach {
          if (margin != getAndroidAttribute(it)) {
            return
          }
        }
        removeAndroidAttribute(SdkConstants.ATTR_LAYOUT_MARGIN)
      }
    }
  }

  override fun handlesPainting(): Boolean = true

  override fun createInteraction(screenView: ScreenView,
                                 x: Int,
                                 y: Int,
                                 component: NlComponent) = SceneInteraction(screenView)

  override fun createTargets(sceneComponent: SceneComponent): List<Target> {
    val listBuilder = ImmutableList.Builder<Target>()
    RelativeParentTarget.Type.values().forEach { listBuilder.add(RelativeParentTarget(it)) }
    AnchorTarget.Type.values()
      .filterNot { it == AnchorTarget.Type.BASELINE }
      .forEach { listBuilder.add(RelativeAnchorTarget(it, true)) }
    return listBuilder.build()
  }

  override fun createChildTargets(parentComponent: SceneComponent, childComponent: SceneComponent): List<Target> {
    val listBuilder = ImmutableList.builder<Target>()
    listBuilder.add(RelativeDragTarget())

    RESIZE_TARGETS.forEach { listBuilder.add(RelativeResizeTarget(it)) }
    AnchorTarget.Type.values()
      .filterNot { it == AnchorTarget.Type.BASELINE }
      .forEach { listBuilder.add(RelativeAnchorTarget(it, false)) }
    RelativeWidgetTarget.Type.values()
      .filter { it !== RelativeWidgetTarget.Type.BASELINE || childComponent.nlComponent.getBaseline() != -1 }
      .forEach { listBuilder.add(RelativeWidgetTarget(it)) }
    return listBuilder.build()
  }

  override fun addToolbarActions(actions: MutableList<ViewAction>) {
    actions.add(getViewOptionsAction())
    actions.addAll(getToggleSizeActions())
    actions.add(ToggleAutoConnectAction())
  }

  override fun getPlaceholders(sceneComponent: SceneComponent, draggedComponents: List<SceneComponent>) =
    listOf(RelativePlaceholder(sceneComponent))
}

private val RESIZE_TARGETS = listOf(
  ResizeBaseTarget.Type.LEFT_TOP,
  ResizeBaseTarget.Type.LEFT_BOTTOM,
  ResizeBaseTarget.Type.RIGHT_TOP,
  ResizeBaseTarget.Type.RIGHT_BOTTOM
)

val MARGINS_ATTRS = arrayOf(
  SdkConstants.ATTR_LAYOUT_MARGIN_START,
  SdkConstants.ATTR_LAYOUT_MARGIN_TOP,
  SdkConstants.ATTR_LAYOUT_MARGIN_END,
  SdkConstants.ATTR_LAYOUT_MARGIN_BOTTOM
)
